### PR TITLE
010 infotech

### PR DESCRIPTION
**Pull request by vEpiphyte**
- Bugfix in snap prop set callbacks
- Bugfix in subs population on a node
- Bugfix in FilePath subs population
- Initial port of infotech model

### CODE REVIEW

File: synapse/lib/snap.py [hunk 1]
```diff
@@ -304,7 +304,7 @@ def _addNodeFnib(self, fnib, props=None):
         form.wasAdded(node)
 
         # now we must fire all his prop sets
-        for name, valu in node.props.items():
```
Comment (vEpiphyte) [inline synapse/lib/snap.py]: callbacks can cause this dict to change in size, throwing a runtimeerror

File: synapse/models/infotech.py [hunk 2]
```diff
@@ -1,16 +1,16 @@
 import logging
 
+import synapse.exc as s_exc
 import synapse.common as s_common
-import synapse.datamodel as s_datamodel
-
-import synapse.lib.version as s_version
+# import synapse.datamodel as s_datamodel
```
Comment (vEpiphyte) [inline synapse/models/infotech.py]: dead code

File: synapse/tests/test_model_infotech.py [hunk 3]
```diff
@@ -293,127 +316,250 @@ def test_model_infotech_semvertype(self):
...
+                self.eq(node.get('file'), fbyts)
+                self.eq(node.get('ctime'), tick)
+                self.eq(node.get('mtime'), tick + 1)
+                self.eq(node.get('atime'), tick + 2)
+                self.eq(node.get('path'), fpath)
+                self.eq(node.get('path:dir'), 'c:/temp')
+                self.eq(node.get('path:base'), 'yourfiles.rar')
+                self.eq(node.get('path:ext'), 'rar')
+
+                # rprops = {
```
Comment (vEpiphyte) [inline synapse/tests/test_model_infotech.py]: test to fix

File: synapse/lib/node.py [hunk 4]
```diff
@@ -131,7 +131,7 @@ def set(self, name, valu, init=False):
 
             for subname, subvalu in subs.items():
 
-                full = prop.name + ':' + name
+                full = prop.name + ':' + subname
```
[the remark this answers is not in the document]
Comment (vertexmc) [reply synapse/lib/node.py]: :+1:

File: synapse/models/infotech.py [hunk 5]
```diff
@@ -25,559 +24,770 @@ class SemverType(DataType):
...
+            if subs is None:
+                raise s_common.BadTypeValu(valu=valu,
+                                           mesg='Unable to brute force version parts out of the string')
+            if subs:
+                valu = s_version.packVersion(subs.get('major'),
+                                             subs.get('minor', 0),
+                                             subs.get('patch', 0))
+                return valu, subs
+
+    def _onFormItDevStr(self, node):
```
Comment (vertexmc) [inline synapse/models/infotech.py]: Do these two functions also need to have oldv in their args?

File: synapse/models/infotech.py [hunk 6]
```diff
@@ -25,559 +24,770 @@ class SemverType(DataType):
...
+                ('it:hostname', ('str', {'strip': True, 'lower': True}), {
+                    'doc': 'The name of a host or sytsem',
+                }),
+                ('it:host', ('guid', {}), {
+                    'doc': 'A GUID that represents a host or system.'
+                }),
+                ('it:hosturl', ('comp', {'fields': (('host', 'it:host'), ('url', 'inet:url'))}), {
+                    'doc': 'A url hosted on or served by a host or system.',
+                }),
+                # TODO We probably want a cve to be linked to a softver via comp type
```
Comment (vertexmc) [inline synapse/models/infotech.py]: Agreed

File: synapse/models/infotech.py [hunk 7]
```diff
@@ -25,559 +24,770 @@ class SemverType(DataType):
...
+                    'doc': 'A arbitrary, unversioned software product.',
+                }),
+                ('it:prod:softver', ('guid', {}), {
+                    'doc': 'A version of a particular software product.'
+                }),
+                ('it:hostsoft', ('comp', {'fields': (('host', 'it:host'), ('softver', 'it:prod:softver'))}), {
+                   'doc': 'A version of a software product which is present on a given host.',
+                }),
+                ('it:av:sig', ('comp', {'fields': (('soft', 'it:prod:soft'), ('name', ('str', {'lower': True})))}), {
+                   'doc': 'A signature name A signature name within the namespace of an antivirus engine name.'
```
Comment (vertexmc) [inline synapse/models/infotech.py]: Docstring typo

File: synapse/models/infotech.py [hunk 8]
```diff
@@ -25,559 +24,770 @@ class SemverType(DataType):
...
+                ('it:dev:pipe', ('str', {}), {
+                    'doc': 'A string representing a named pipe.',
+                }),
+                ('it:dev:mutex', ('str', {}), {
+                    'doc': 'A string representing a mutex.',
+                }),
+                ('it:dev:int', ('int', {}), {
+                    'doc': 'A developer selected integer constant.',
+                }),
+                ('it:dev:regkey', ('str', {}), {
```
Comment (vertexmc) [inline synapse/models/infotech.py]: This used to be a sub of `it:dev:str`

File: synapse/models/infotech.py [hunk 9]
```diff
@@ -25,559 +24,770 @@ class SemverType(DataType):
...
-                    'subof': 'comp',
-                    'fields': 'host,it:host|softver,it:prod:softver',
-                    'doc': 'A version of a software product which is present on a given host.'}),
-
-                ('it:auth:passwdhash', {
-                    'subof': 'guid',
-                    'ex': '(hash:md5=17d3533fba2669f84a225a9a04caa783)'}),
+                    'ex': 'HKEY_LOCAL_MACHINE\\SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Run',
+                }),
+                ('it:dev:regval', ('guid', {}), {
```
Comment (vertexmc) [inline synapse/models/infotech.py]: This used to be a comp of `it:dev:regkey`, should it be a sub of`it:dev:regkey`?

File: synapse/models/infotech.py [hunk 10]
```diff
@@ -25,559 +24,770 @@ class SemverType(DataType):
...
-                        'doc': 'The file that triggered the signature hit.'}),
-                    ('sig', {'ptype': 'it:av:sig',
-                        'doc': 'The signature that the file triggered on.'}),
-                ]),
-
+                ('it:host', {}, (
+                    ('name', ('it:hostname', {}), {
+                        'doc': 'The name of the host or system.',
+                    }),
+                    ('desc', ('str', {}), {
```
Comment (vertexmc) [inline synapse/models/infotech.py]: Do we need to do anything about `str:txt`s becoming `str`? I've been marking the ones I see

File: synapse/models/infotech.py [hunk 11]
```diff
@@ -25,559 +24,770 @@ class SemverType(DataType):
...
+                    }),
+                    #  FIXME we probably eventually need a bunch of stuff here...
+                    ('ipv4', ('inet:ipv4', {}), {
+                        'doc': 'The last known ipv4 address for the host.'
+                    }),
+                    ('latlong', ('geo:latlong', {}), {
+                        'doc': 'The last known location for the host.'
+                    }),
+                )),
+                ('it:hosturl', {}, (
```
Comment (vertexmc) [inline synapse/models/infotech.py]: Weird that there was no existing form for this. Probably should have been.

File: synapse/models/infotech.py [hunk 12]
```diff
@@ -25,559 +24,770 @@ class SemverType(DataType):
...
+                    ('host', ('it:host', {}), {
+                        'ro': True,
```
Comment (vertexmc) [inline synapse/models/infotech.py]: Do we need `'ro': True` anymore?

File: synapse/models/infotech.py (hunk 5 above)
Comment (vEpiphyte) [reply synapse/models/infotech.py]: onForm callbacks online get the node

File: synapse/models/infotech.py [hunk 13]
```diff
@@ -25,559 +24,770 @@ class SemverType(DataType):
...
-                    ('time', {'ptype': 'time', 'doc': 'The time the port was bound.'}),
-                    ('port', {'ptype': 'inet:port', 'doc': 'The bound (listening) UDP port.'}),
-                    ('ipv4', {'ptype': 'inet:ipv4', 'doc': 'The IPv4 specified to bind().'}),
-                    ('ipv6', {'ptype': 'inet:ipv6', 'doc': 'The IPv6 specified to bind().'}),
+                ('it:auth:passwdhash', {}, (
+                    ('salt', ('hex', {}), {
+                        'doc': 'The (optional) hex encoded salt value used to calculate the password hash.',
+                    }),
+                    ('hash:md5', ('hash:md5', {}), {
+                        'doc': 'The MD5 password hash value.',
```
Comment (vertexmc) [inline synapse/models/infotech.py]: Nice docstring catch

File: synapse/models/infotech.py [hunk 14]
```diff
@@ -25,559 +24,770 @@ class SemverType(DataType):
...
+                    ('host', ('it:host', {}), {
+                        'doc': 'The host running the process that created the named pipe. Typically the same host referenced in :proc, if present.',
+                    }),
+                    ('exe', ('file:bytes', {}), {
+                        'doc': 'The specific file containing code that created the named pipe. May or may not be the same :exe specified in :proc, if present.',
+                    }),
+                    ('time', ('time', {}), {
+                        'doc': 'The time the mutex was created.',
+                    }),
+                    ('name', ('it:dev:mutex', {}), {
```
Comment (vertexmc) [inline synapse/models/infotech.py]: Should this be an `it:dev:pipe`? it was before

Comment (vEpiphyte) [reply synapse/models/infotech.py]: good catch

File: synapse/models/infotech.py [hunk 15]
```diff
@@ -25,559 +24,770 @@ class SemverType(DataType):
...
+                    }),
+                    ('time', ('time', {}), {
+                        'doc': 'The time the URL was requested.',
+                    }),
+                    ('url', ('inet:url', {}), {
+                        'doc': 'The URL that was requested.',
+                    }),
+                    # FIXME Needs inet:addr
+                    # ('client', ('inet:addr', {}), {
+                    #     'doc': 'The address of hte client during the URL retrieval.'
```
Comment (vEpiphyte) [inline synapse/models/infotech.py]: typo

File: synapse/models/infotech.py [hunk 16]
```diff
@@ -25,559 +24,770 @@ class SemverType(DataType):
...
+                    ('exe', ('file:bytes', {}), {
+                        'doc': 'The specific file containing code that requested the URL. May or may not be the same :exe specified in :proc, if present.',
+                    }),
+                    ('time', ('time', {}), {
+                        'doc': 'The time the URL was requested.',
+                    }),
+                    ('url', ('inet:url', {}), {
+                        'doc': 'The URL that was requested.',
+                    }),
+                    # FIXME Needs inet:addr
```
Comment (vertexmc) [inline synapse/models/infotech.py]: Yep - should have this shortly.

File: synapse/tests/test_model_infotech.py [hunk 17]
```diff
@@ -1,277 +1,300 @@
...
+# Test it:prod:softver by range!
+#
+
+class InfotechModelTest(s_test.SynTest):
+    def test_it_forms_simple(self):
+        with self.getTestCore() as core:
+            with core.snap(write=True) as snap:
+                node = snap.addNode('it:hostname', 'Bobs Computer')
+                self.eq(node.ndef[1], 'bobs computer')
+                # FIXME add test for it:host
```
Comment (vEpiphyte) [inline synapse/tests/test_model_infotech.py]: can remove comment

File: synapse/models/infotech.py [hunk 18]
```diff
@@ -25,559 +24,770 @@ class SemverType(DataType):
...
-                         'doc': 'The host running the process that wrote to the file. Typically the same host referenced in :proc, if present.'}),
-                    ('exe', {'ptype': 'file:bytes',
-                         'doc': 'The specific file containing code that wrote to the file. May or may not be the same :exe referenced in :proc, if present.'}),
-                    ('time', {'ptype': 'time', 'doc': 'The time the file was written to / modified.'}),
-                    ('path', {'ptype': 'file:path', 'doc': 'The path where the file was modified.'}),
-                    ('path:dir', {'ptype': 'file:path', 'doc': 'The parent directory of the file path (parsed from :path).', 'ro': 1}),
-                    ('path:ext', {'ptype': 'str:lwr', 'doc': 'The file extension of the file name (parsed from :path).', 'ro': 1}),
-                    ('path:base', {'ptype': 'file:base', 'doc': 'The final component of the file path (parsed from :path).', 'ro': 1}),
-                    ('file', {'ptype': 'file:bytes', 'doc': 'The file that was modified.'}),
+                ('it:exec:bind', {}, (
```
Comment (vertexmc) [inline synapse/models/infotech.py]: :-D nice that we can get rid of bind:tcp, etc

File: synapse/tests/test_model_infotech.py [hunk 19]
```diff
@@ -1,277 +1,300 @@
...
+                    'vers': 'V1.0.1-beta+exp.sha.5114f85',
+                    'url': url1,
+                    'software': prod0,
+                    'arch': 'amd64'
+                }
+                ver0 = s_common.guid()
+                node = snap.addNode('it:prod:softver', ver0, vprops)
+
+                self.eq(node.ndef[1], ver0)
+                self.eq(node.get('software'), prod0)
```
Comment (vEpiphyte) [inline synapse/tests/test_model_infotech.py]: need to confirm arch is set

File: synapse/models/infotech.py [hunk 20]
```diff
@@ -25,559 +24,770 @@ class SemverType(DataType):
...
-                    ('hash:sha1', {'ptype': 'hash:sha1', 'ro': 1,
-                        'doc': 'The SHA1 password hash value.'}),
-
-                    ('hash:sha256', {'ptype': 'hash:sha256', 'ro': 1,
-                        'doc': 'The SHA256 password hash value.'}),
-
-                    ('hash:sha512', {'ptype': 'hash:sha512', 'ro': 1,
-                        'doc': 'The SHA512 password hash value.'}),
-
-
```
Comment (vertexmc) [inline synapse/models/infotech.py]: extra .

File: synapse/models/infotech.py (hunk 9 above)
Comment (vertexmc) [reply synapse/models/infotech.py]: Nevermind, didn't see the optfields

File: synapse/lib/types.py [hunk 21]
```diff
@@ -622,11 +622,39 @@ def indxByEq(self, valu):
...
+    '''
+    def __init__(self, modl, fields):
+        collections.defaultdict.__init__(self)
+        self.modl = modl
+        self.fields = {name: tname for name, tname in fields}
+
+    def __missing__(self, key):
+        val = self.fields.get(key)
+        if not val:
+            raise s_exc.BadTypeDef(valu=key, mesg='unconfigured field requested')
```
Comment (vertexmc) [inline synapse/lib/types.py]: not covered

File: synapse/lib/types.py [hunk 22]
```diff
@@ -622,11 +622,39 @@ def indxByEq(self, valu):
...
+        if isinstance(val, str):
+            _type = self.modl.type(val)
+            if not _type:
+                raise s_exc.BadTypeDef(valu=val, mesg='type is not present in datamodel')
```
Comment (vertexmc) [inline synapse/lib/types.py]: not covered

File: synapse/lib/types.py [hunk 23]
```diff
@@ -622,11 +622,39 @@ def indxByEq(self, valu):
...
+        else:
+            # val is a type, opts pair
+            tname, opts = val
+            basetype = self.modl.type(tname)
+            if not basetype:
+                raise s_exc.BadTypeDef(valu=val, mesg='type is not present in datamodel')
```
Comment (vertexmc) [inline synapse/lib/types.py]: not covered

File: synapse/models/infotech.py (hunk 7 above)
Comment (vEpiphyte) [reply synapse/models/infotech.py]: fixed

File: synapse/models/infotech.py (hunk 15 above)
Comment (vEpiphyte) [reply synapse/models/infotech.py]: fixed

File: synapse/tests/test_model_infotech.py (hunk 17 above)
Comment (vEpiphyte) [reply synapse/tests/test_model_infotech.py]: fixed

File: synapse/models/infotech.py (hunk 18 above)
Comment (vEpiphyte) [reply synapse/models/infotech.py]: :)

File: synapse/tests/test_model_infotech.py (hunk 19 above)
Comment (vEpiphyte) [reply synapse/tests/test_model_infotech.py]: fixed

File: synapse/models/infotech.py (hunk 20 above)
Comment (vEpiphyte) [reply synapse/models/infotech.py]: where is this????????????

Comment (vEpiphyte) [reply synapse/models/infotech.py]: found it

File: synapse/lib/types.py (hunk 21 above)
Comment (vEpiphyte) [reply synapse/lib/types.py]: that's why its left as a fixme :(

File: synapse/lib/types.py (hunk 22 above)
Comment (vEpiphyte) [reply synapse/lib/types.py]: that's why its left as a fixme :(

File: synapse/lib/types.py (hunk 23 above)
Comment (vEpiphyte) [reply synapse/lib/types.py]: that's why its left as a fixme :(

**Pull request merged**